src/weights.jl

```diff
@@ -5,15 +5,18 @@ abstract type AbstractWeights{S<:Real, T<:Real, V<:AbstractVector{T}} <: Abstrac
     @weights name
 
 Generates a new generic weight type with specified `name`, which subtypes `AbstractWeights`
-and stores the `values` (`V<:RealVector`) and `sum` (`S<:Real`).
+and stores the `values` (`V<:RealVector` using 1-based indexing) and `sum` (`S<:Real`).
 """
 macro weights(name)
     return quote
         mutable struct $name{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
             values::V
             sum::S
         end
-        $(esc(name))(values::AbstractVector{<:Real}) = $(esc(name))(values, sum(values))
+        function $(esc(name))(values::AbstractVector{<:Real})
+            Base.require_one_based_indexing(values)
```

Comment:
This requires Julia >= 1.2: https://github.com/JuliaLang/julia/blob/8a0a7199ddc45ec70d8a0b74ab397ca69d34b50e/base/abstractarray.jl#L119-L120